### ISSUE TITLE
Singular fails to build when Flint is not in a standard location

### ISSUE THREAD

**Issue by wbhart** [Singular/Sources#971]
@fingolfin requested we add a ticket for this issue, so he can help track it down.

When building Singular with the --with-flint option with Flint in a non-standard location, the build fails, i.e. if you build Singular with the --with-flint option and specify a directory where Flint is located that Singular doesn't already look for some other reason, the build fails. Essentially it ignores whatever directory you pass via the --with-flint option.

I believe this has been happening ever since https://github.com/Singular/Sources/blob/spielwiese/libpolys/coeffs/flintcf_Qrat.cc was added to the Singular build system (not when the file itself was initially added to the repo; it wasn't added to the build system at that time), and not before.

We have verified that the AM variable that specifies this directory is empty when linking occurs in the part of the build that deals with https://github.com/Singular/Sources/blob/spielwiese/libpolys/coeffs/flintcf_Qrat.cc.

For some reason that AM variable is empty, but should contain the directory passed at configure time via the --with-flint option. The result is Singular tries to link against Flint at that point, but the directory where the Flint library is located is simply missing from the linker invocation (during the build process).

**Comment by wbhart** [comment]
I am currently verifying whether the latest changes made by Hans have changed the status of this issue. I will report back when the build either fails or completes, assuming I am not interrupted.

**Comment by fingolfin** [comment]
I just built Singular `master` using `--with-flint=PATH` against a non-standard flint, with no flint available anywhere else Flint might pick up, but this worked without a hitch. Perhaps that means @hannes14 fixed it?

If not (i.e., if @wbhart can still reproduce it), some more info would be nice?

- on which OS does the problem appear? (I am on macOS 10.14)
- is the custom flint the only one, or is there also a different flint somewhere Singular might "see" it by accident?
- are any other `configure` options specified?
- what is the exact error message?
- the full output of `configure` and the `config.log` of the top `configure`, and of `libpolys/configure`, each

**Comment by wbhart** [comment]
I can reproduce *an* issue, but not quite what I remember. In fact it is complaining about an include (of flint/flint.h), not linking against the Flint library at the moment. Hans has made another fix which may resolve that. I will report back when that either succeeds or fails.

I am using WSL stuck half way between Ubuntu 16.04 and 18.04 due to the Ubuntu upgrader for the WSL being broken at the time I upgraded. However, the original problem was also reproducible on Gentoo (version unknown).

More in another 15 minutes.

**Comment by wbhart** [comment]
Indeed, the latest fixes (including the one today), fix the problem. (That is, after I worked around numerous other problems I encountered along the way.)

Hans summarised the solution something like this: Singular was only checking the last specified place for libflint. This was usually fine since on all our machines this was usually the default location. In my case, that was not so. Therefore it didn't find Flint, silently failed (by design) and set the directory to be empty.

Anyhow, I'm glad that is finally resolved!
